Build bcachefs-tools with podman for gokrazy

peebs opened the pull request:
I never figured out how to cross-compile bcachefs-tools directly but I got the bcachefs equivalent to `gokrazy/mkfs` working using `podman build --platform=ARCH` instead. Open for feedback on approach here!

Also, when executing `bcachefs format /dev/...` in mkfs.go it seems `format` is somehow getting passed as the first argument and I don't understand why.  I just call `bcachefs /dev/...` and the `format` subcommand is run on gokrazy.

stapelberg commented on Dockerfile:
```diff
@@ -0,0 +1,20 @@
+ENV SOURCE_DATE_EPOCH 1600000000
+
+# build bcachefs-tools
+RUN mkdir /src /bins && wget https://github.com/koverstreet/bcachefs-tools/archive/refs/heads/master.zip && unzip -d /src /master.zip && mv /src/bcachefs-tools-master/* /src
```
Can you change this to use a tagged version instead of the master branch? With e2fsprogs, we also track a released version.

peebs replied:
Done

stapelberg commented on mkfs.go:
```diff
@@ -0,0 +1,136 @@
+	if err := makeFS(); err != nil {
+		fmt.Fprintf(os.Stderr, "%v\n", err)
+	}
+	/*
```
Why is this commented out?

peebs replied:
Left over debugging, updated.

stapelberg commented on mkfs.go:
```diff
@@ -0,0 +1,136 @@
+	if err != nil {
+		return err
+	}
+	//defer os.RemoveAll(tmp)
```
Why is this commented out?

peebs replied:
Left over debugging, updated.

stapelberg reviewed (changes requested):
Thanks for your PR!

stapelberg commented:
> I never figured out how to cross-compile bcachefs-tools directly but I got the bcachefs equivalent to `gokrazy/mkfs` working using `podman build --platform=ARCH` instead. Open for feedback on approach here!

I don’t mind the `--platform` approach (but haven’t actually tested yet whether it works on my machine).

I’m not familiar with cross-compiling Rust, so I’m not sure either if there is a better approach.

> Also, when executing `bcachefs format /dev/...` in mkfs.go it seems `format` is somehow getting passed as the first argument and I don't understand why. I just call `bcachefs /dev/...` and the `format` subcommand is run on gokrazy.

Not sure I follow? At first I thought maybe you’re running into [the odd execve(2) API](https://man.zekjur.net/bookworm/manpages-dev/execve.2.en.html), where the program name needs to be specified both as the first parameter, and *again* in the argv array. But maybe you mean something else entirely?

peebs commented:
> I don’t mind the `--platform` approach (but haven’t actually tested yet whether it works on my machine).
> 
Ok, happy to change things so its easy to build for you if necessary! Podman build just worked for me on Fedora, I could try the Docker equivalent too.  
 
> Not sure I follow? At first I thought maybe you’re running into [the odd execve(2) API](https://man.zekjur.net/bookworm/manpages-dev/execve.2.en.html), where the program name needs to be specified both as the first parameter, and _again_ in the argv array. But maybe you mean something else entirely?

My quandary is that `bcachefs format /dev/...` is the command to format a block device. When I test the `bcachefs` binary locally this is confirmed.  

However when I am exec'ing `bcachefs` on gokrazy via mkfs.go I am forced to drop the `format` sub-command. If I just call the binary without `format` it will ask for a block device as the next argument. If I call it with `format` it will tell me `format` isn't a block device. So I dropped `format` argument here and it worked:
```
	// exec bcachefs format on rootdev partition
	mkfs := exec.Command(filepath.Join(tmp, "ld-linux-aarch64.so.1"), filepath.Join(tmp, "bcachefs"), dev)
	mkfs.Env = append(os.Environ(), "LD_LIBRARY_PATH="+tmp)
	mkfs.Stdout = os.Stdout
	mkfs.Stderr = os.Stderr
	log.Printf("exec bcachefs with args: %v", mkfs.Args)
	if err := mkfs.Run(); err != nil {
		return fmt.Errorf("%v: %v", mkfs.Args, err)
	}
```

stapelberg commented on go.mod:
```diff
@@ -0,0 +1,7 @@
+module github.com/gokrazy/mkfs
```
Update this to github.com/gokrazy/mkfs.bcachefs, otherwise `gok add` will add the wrong path to config.json :)

stapelberg commented on Makefile:
```diff
@@ -0,0 +1,23 @@
+# Using podman build --platform to cross compile. Podman will run non-native
+# platforms via `qemu-user-static` by default.
+
+all: third_party/bcachefs-tools/amd64/bcachefs third_party/bcachefs-tools/arm64/bcachefs third_party/bcachefs-tools/arm/bcachefs 
```
Here and in *.go, please include the version in the path, i.e. `third_party/bcachefs-tools-1.7.0`

stapelberg commented on README.md:
````diff
@@ -0,0 +1,29 @@
+
+```
+git clone https://github.com/gokrazy/mkfs.bcachefs
+cd mkfs
````
cd mkfs.bcachefs

stapelberg commented on mkfs.go:
```diff
@@ -0,0 +1,134 @@
+}
+
+func main() {
+	if err := checkExistingFS(); err != nil {
```
You have restructured the program such that it also reboots even when not creating a file system. This makes it impossible to include the mkfs.bcachefs package in gokrazy installations. `gok run` still works, but adding the package permanently should also work.

(Also, it tries to create the file system even when one is already mounted.)

Can you revert your changes to mkfs.go as much as possible please? That will also make ongoing maintenance easier.

stapelberg reviewed (commented):
I tested it and it worked! A couple of comments before we can merge it, though. Thanks!

stapelberg commented:
> My quandary is that `bcachefs format /dev/...` is the command to format a block device. When I test the `bcachefs` binary locally this is confirmed.
> 
> However when I am exec'ing `bcachefs` on gokrazy via mkfs.go I am forced to drop the `format` sub-command. If I just call the binary without `format` it will ask for a block device as the next argument. If I call it with `format` it will tell me `format` isn't a block device. So I dropped `format` argument here and it worked:

I figured it out: the bcachefs tool checks argv[0] to see whether it was symlinked as mkfs.bcachefs or similar: https://github.com/koverstreet/bcachefs-tools/blob/0d570e822b350320a2787033b0b92c49e579e2a8/src/bcachefs.rs#L77-L78

Because we extract bcachefs to /tmp/gokrazy-mkfs.bcachefs-xxxxxx, this code path triggers and the different argument syntax applies.

We can leave this as-is and add a comment for explanation, or change the path.

stapelberg commented on README.md:
````diff
@@ -0,0 +1,29 @@
+```
+
+## Updating
+Install podman and:
````
Ah, one more thing: On Arch Linux, I needed not just podman, but also qemu-user-static-binfmt. Please mention it here as well.

peebs committed:
copy gokrazy/mkfs at commit b6755f9

peebs committed:
build bcachefs-tools v1.7.0 and format SD card

peebs commented:
> I figured it out: the bcachefs tool checks argv[0] to see whether it was symlinked as mkfs.bcachefs or similar: https://github.com/koverstreet/bcachefs-tools/blob/0d570e822b350320a2787033b0b92c49e579e2a8/src/bcachefs.rs#L77-L78
> 
> Because we extract bcachefs to /tmp/gokrazy-mkfs.bcachefs-xxxxxx, this code path triggers and the different argument syntax applies.
> 
> We can leave this as-is and add a comment for explanation, or change the path.

I changed the path in case anyone working on this gets confused about it too. Also addressed all the comments and reverted extra `mkfs.go` changes. All tested and working again for me. Thanks for the review.

stapelberg commented:
Thanks!
